mungegithub/mungers/milestone-maintainer.go

```diff
@@ -0,0 +1,461 @@
+	}
+
+	milestoneRemovedLabel          = strings.ToLower(strings.Replace(milestoneRemoved, " ", "-", -1))
+	milestoneLabelsIncompleteLabel = strings.ToLower(strings.Replace(milestoneLabelsIncomplete, " ", "-", -1))
+	milestoneLabelsCompleteLabel   = strings.ToLower(strings.Replace(milestoneLabelsComplete, " ", "-", -1))
+	milestoneLabels                = []string{milestoneRemovedLabel, milestoneLabelsCompleteLabel, milestoneLabelsIncompleteLabel}
+)
+
+// milestoneNotification configures the active notification for the munger
+type milestoneNotification struct {
+	deleteExisting bool
+	description    string
+	message        string
+}
+
+// MilestoneMaintainer enforces the process for sheperding issues into the release.
+type MilestoneMaintainer struct {
+	botName  string
+	features *features.Features
+
+	activeMilestone string
+}
+
+func init() {
+	RegisterMungerOrDie(&MilestoneMaintainer{})
+}
+
+// Name is the name usable in --pr-mungers
+func (m *MilestoneMaintainer) Name() string { return "milestone-maintainer" }
+
+// RequiredFeatures is a slice of 'features' that must be provided
+func (m *MilestoneMaintainer) RequiredFeatures() []string { return []string{} }
+
+// Initialize will initialize the munger
+func (m *MilestoneMaintainer) Initialize(config *github.Config, features *features.Features) error {
+	if len(m.activeMilestone) == 0 {
+		return errors.New("active-milestone must be supplied")
+	}
+	m.botName = config.BotName
+	m.features = features
+	return nil
+}
+
+// EachLoop is called at the start of every munge loop
+func (m *MilestoneMaintainer) EachLoop() error { return nil }
+
+// RegisterOptions registers options for this munger; returns any that require a restart when changed.
+func (m *MilestoneMaintainer) RegisterOptions(opts *options.Options) sets.String {
+	opts.RegisterString(&m.activeMilestone, "active-milestone", "", "The active milestone that this munger will maintain issues for.")
+	return sets.NewString("active-milestone")
+}
+
+// Munge is the workhorse the will actually make updates to the issue
+func (m *MilestoneMaintainer) Munge(obj *github.MungeObject) {
+	if m.ignoreObject(obj) {
+		return
+	}
+
+	comment, ok := latestNotificationComment(obj, m.botName)
+	if !ok {
+		return
+	}
+
+	labelName, notifyState := notificationState(obj, comment, m.botName)
+	// Always attempt to ensure the milestone label
+	if len(labelName) > 0 && !ensureLabel(obj, labelName) {
+		return
+	}
+	if notifyState == nil {
+		// No notification change is required
+		return
+	}
+
+	if notifyState.deleteExisting {
+		if err := obj.DeleteComment(comment.Source.(*githubapi.IssueComment)); err != nil {
+			return
+		}
+	}
+
+	notification := c.Notification{
+		Name:      milestoneNotifierName,
+		Arguments: notifyState.description,
+		Context:   notifyState.message,
+	}
+	if err := notification.Post(obj); err != nil {
+		return
+	}
+
+	if labelName != milestoneRemovedLabel {
```

Comment:
nit: considering putting the ClearMilestone in the if block, and check ==.  IMHO it reads a little more clearly to a first time reader that way (if labelName == milestoneRemovedLabel then ClearMilestone, vs if labelName != milestoneRemovedLabel then return; remove milestone.) - I realize this breaks convention with the previous statements in this function, so could be convinced to leave it as is.